core/src/price_estimation/orderbook_based.rs

```diff
@@ -0,0 +1,144 @@
+use super::price_source::PriceSource;
+use crate::models::{BatchId, TokenId};
+use crate::orderbook::StableXOrderBookReading;
+use anyhow::Result;
+use futures::future::{BoxFuture, FutureExt as _};
+use pricegraph::{Orderbook, TokenPair};
+use primitive_types::U256;
+use std::collections::HashMap;
+use std::time::SystemTime;
+
+pub struct PricegraphEstimator<'a> {
+    orderbook_reader: &'a (dyn StableXOrderBookReading),
+}
+
+const ONE_OWL: f64 = 1_000_000_000_000_000_000.0;
+
+impl<'a> PriceSource for PricegraphEstimator<'a> {
+    fn get_prices<'b>(
+        &'b self,
+        tokens: &'b [TokenId],
+    ) -> BoxFuture<'b, Result<HashMap<TokenId, u128>>> {
+        async move {
+            let batch = BatchId::currently_being_solved(SystemTime::now())?;
+            let (account_state, orders) =
+                self.orderbook_reader.get_auction_data(batch.into()).await?;
+            let mut orderbook = Orderbook::from_elements(orders.iter().map(|order| {
+                order.to_element(U256(
+                    account_state
+                        .read_balance(order.sell_token, order.account_id)
+                        .0,
+                ))
+            }));
+            orderbook.reduce_overlapping_orders();
+            Ok(self.estimate_prices(tokens, orderbook))
+        }
+        .boxed()
+    }
+}
+
+// Private trait to facilitate testing this module
+#[cfg_attr(test, mockall::automock)]
+trait Pricegraph {
+    fn estimate_price(&mut self, pair: TokenPair, volume: f64) -> Option<f64>;
+}
+
+impl Pricegraph for Orderbook {
+    fn estimate_price(&mut self, pair: TokenPair, volume: f64) -> Option<f64> {
+        self.fill_market_order(pair, volume)
+    }
+}
+
+impl<'a> PricegraphEstimator<'a> {
+    #[allow(dead_code)]
+    pub fn new(orderbook_reader: &'a (dyn StableXOrderBookReading)) -> Self {
+        Self { orderbook_reader }
+    }
+
+    fn estimate_prices(
+        &self,
+        tokens: &[TokenId],
+        mut pricegraph: impl Pricegraph,
+    ) -> HashMap<TokenId, u128> {
+        tokens
+            .iter()
+            .flat_map(|token| {
+                let price = if token == &TokenId::reference() {
+                    1.0
+                } else {
+                    let pair = TokenPair {
+                        buy: token.0,
+                        sell: TokenId::reference().0,
+                    };
+                    pricegraph.estimate_price(pair, ONE_OWL)?
```

Comment:
@nlordell do you think we should be cloning the price graph here so that we don't "eat" up the liquidity in between estimates?

Reply:
Yes. I think the pricegraph needs a new API that uses the orderbook graph internally.